src/rust/cryptography-x509-validation/src/certificate.rs

```diff
@@ -0,0 +1,116 @@
+// This file is dual licensed under the terms of the Apache License, Version
+// 2.0, and the BSD License. See the LICENSE file in the root of this repository
+// for complete details.
+
+use asn1::ObjectIdentifier;
+use cryptography_x509::{
+    certificate::Certificate,
+    extensions::{Extension, Extensions},
+    oid::{
+        AUTHORITY_KEY_IDENTIFIER_OID, BASIC_CONSTRAINTS_OID, EXTENDED_KEY_USAGE_OID, KEY_USAGE_OID,
+        SUBJECT_ALTERNATIVE_NAME_OID, SUBJECT_KEY_IDENTIFIER_OID,
+    },
+};
+
+use crate::{backend::Backend, extensions::ExtensionError, types::Name};
+
+#[derive(Debug, PartialEq)]
+pub enum CertificateError {
+    DuplicateExtension(asn1::ObjectIdentifier),
+    InvalidExtension(ExtensionError),
+}
+
+impl From<ExtensionError> for CertificateError {
+    fn from(value: ExtensionError) -> Self {
+        CertificateError::InvalidExtension(value)
+    }
+}
+
+pub(crate) trait CertificateExt<'a>: private::Sealed {
+    fn is_self_issued(&self) -> bool;
+    fn is_self_signed<B: Backend>(&self) -> bool;
+
+    fn issuer(&self) -> &Name<'a>;
+    fn subject(&self) -> &Name<'a>;
+
+    fn extensions(&'a self) -> Result<Extensions<'a>, CertificateError>;
+    fn extension_is_critical(&self, oid: &ObjectIdentifier) -> bool;
+
+    fn authority_key_identifier(&'a self) -> Result<Option<Extension<'a>>, CertificateError>;
+    fn subject_key_identifier(&'a self) -> Result<Option<Extension<'a>>, CertificateError>;
+    fn key_usage(&'a self) -> Result<Option<Extension<'a>>, CertificateError>;
+    fn basic_constraints(&'a self) -> Result<Option<Extension<'a>>, CertificateError>;
+    fn subject_alternative_name(&'a self) -> Result<Option<Extension<'a>>, CertificateError>;
+    fn extended_key_usage(&'a self) -> Result<Option<Extension<'a>>, CertificateError>;
```

Comment:
These seem very low value to me, why not `fn extension<T: Asn1Readable>(&self, oid: ObjectIdentifier) -> Result<T>`?

Reply:
Agreed; these are chaff from early experimenting.

Reply:
b0376e671